ansible_runner/cleanup.py

```diff
@@ -0,0 +1,172 @@
+    )
+
+
+def run_command(cmd):
+    '''Given list cmd, runs command and returns standard out, expecting success'''
+    process = subprocess.run(cmd, capture_output=True)
+    stdout = str(process.stdout, encoding='utf-8')
+    if process.returncode != 0:
+        print('Error running command:')
+        print(' '.join(cmd))
+        print('Stdout:')
+        print(stdout)
+        raise RuntimeError('Error running command')
+    return stdout.strip()
+
+
+def is_alive(dir):
+    pidfile = os.path.join(dir, 'pid')
+
+    try:
+        with open(pidfile, 'r') as f:
+            pid = int(f.readline())
+    except IOError:
+        return False
+
+    try:
+        os.kill(pid, signal.SIG_DFL)
+        return(0)
+    except OSError:
+        return(1)
+
+
+def project_idents(dir):
+    """Give dir, give list of idents that we have artifacts for"""
+    try:
+        return os.listdir(os.path.join(dir, 'artifacts'))
+    except (FileNotFoundError, NotADirectoryError):
+        return []
+
+
+def delete_associated_folders(dir):
+    """Where dir is the private_data_dir for a completed job, this deletes related tmp folders it used"""
+    for ident in project_idents(dir):
+        registry_auth_pattern = f'/tmp/{registry_auth_prefix}{ident}_*'
+        for dir in glob.glob(registry_auth_pattern):
+            changed = cleanup_folder(dir)
+            if changed:
+                print(f'Removed associated registry auth dir {dir}')
+
+
+def cleanup_dirs(pattern, exclude_strings=(), grace_period=GRACE_PERIOD_DEFAULT):
+    ct = 0
+    now_time = datetime.datetime.now()
+    for dir in glob.glob(pattern):
+        if any(str(exclude_string) in dir for exclude_string in exclude_strings):
+            continue
+        if grace_period:
+            st = os.stat(dir)
+            modtime = datetime.datetime.fromtimestamp(st.st_mtime)
+            if modtime > now_time - datetime.timedelta(minutes=grace_period):
+                continue
+        if is_alive(dir):
+            print(f'Excluding running project {dir} from cleanup')
+            continue
+        delete_associated_folders(dir)
+        changed = cleanup_folder(dir)
+        if changed:
+            print(f'Removed directory {dir}')
+            ct += 1
+
+    return ct
+
+
+def cleanup_images(images, runtime='podman'):
+    """Note: docker will just untag while podman will remove layers with same command"""
+    rm_ct = 0
+    for image_tag in images:
+        stdout = run_command([runtime, 'images', '--format="{{.Repository}}:{{.Tag}}"', image_tag])
+        if not stdout:
+            continue
+        for discovered_tag in stdout.split('\n'):
+            stdout = run_command([runtime, 'rmi', discovered_tag.strip().strip('"'), '-f'])
+            rm_ct += stdout.count('Untagged:')
+    return rm_ct
+
+
+def prune_images(runtime='podman'):
+    """Run the prune images command and return changed status"""
+    stdout = run_command([runtime, 'image', 'prune', '-f'])
```

Comment:
@fosterseth in AWX we did something a little different, and I'm not totally sure why. It kind of manually deleted dangling images instead of using the prune command.

https://github.com/ansible/awx/blob/7c9626b0e7c0b7835c8ff42970c6f1740f3b10fe/awx/main/tasks.py#L397

This might have been to fix some podman-specific bug, but I lack the knowledge to reproduce that bug. We could add it here using the same commands as you used. Wouldn't hurt.

Reply:
prune here is probably better -- in awx it looks like we wanted to get proper stdout for the images being removed, so it first grabs a list of images, then removes them. Maybe I wasn't able to get it work properly with prune at the time